dtpc/doc/dtpcclock.1
.\" Automatically generated by Pod::Man 4.11 (Pod::Simple 3.35)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "DTPCCLOCK 1"
.TH DTPCCLOCK 1 "2022-11-03" "perl v5.30.0" "DTPC executables"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
dtpcclock \- DTPC daemon task for managing scheduled events
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBdtpcclock\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBdtpcclock\fR is a background \*(L"daemon\*(R" task that periodically performs
scheduled \s-1DTPC\s0 activities.  It is spawned automatically by \fBdtpcadmin\fR in
response to the 's' command that starts operation of the \s-1DTPC\s0 protocol, and
it is terminated by \fBdtpcadmin\fR in response to an 'x' (\s-1STOP\s0) command.
.PP
Once per second, \fBdtpcclock\fR takes the following action:
.Sp
.RS 4
First it executes all \s-1DTPC\s0 events scheduled to occur at any time up to the
current moment:
.Sp
.RS 4
\&\s-1DTPC\s0 ADUs for which an expected positive acknowledgment has not yet arrived
are retransmitted.
.Sp
Received \s-1DTPC\s0 ADUs whose time to live has elapsed are deleted.
.RE
.RE
.RS 4
.Sp
Then \fBdtpcclock\fR increases the ages of all \s-1DTPC\s0 ADUs pending transmission
and initiates transmission of each such \s-1ADU\s0 whose age now equals or exceeds
its aggregation time limit.
.RE
.SH "EXIT STATUS"
.IX Header "EXIT STATUS"
.ie n .IP "0" 4
.el .IP "\f(CW0\fR" 4
.IX Item "0"
\&\fBdtpcclock\fR terminated, for reasons noted in the \fBion.log\fR file.  If this
termination was not commanded, investigate and solve the problem identified
in the log file and use \fBdtpcadmin\fR to restart \fBdtpcclock\fR.
.ie n .IP "1" 4
.el .IP "\f(CW1\fR" 4
.IX Item "1"
\&\fBdtpcclock\fR was unable to attach to \s-1DTPC\s0 protocol operations, probably because
\&\fBdtpcadmin\fR has not yet been run.
.SH "FILES"
.IX Header "FILES"
No configuration files are needed.
.SH "ENVIRONMENT"
.IX Header "ENVIRONMENT"
No environment variables apply.
.SH "DIAGNOSTICS"
.IX Header "DIAGNOSTICS"
The following diagnostics may be issued to the \fBion.log\fR log file:
.IP "dtpcclock can't initialize \s-1DTPC.\s0" 4
.IX Item "dtpcclock can't initialize DTPC."
\&\fBdtpcadmin\fR has not yet initialized \s-1DTPC\s0 protocol operations.
.IP "Can't send finished adu." 4
.IX Item "Can't send finished adu."
An unrecoverable database error was encountered.  \fBdtpcclock\fR terminates.
.IP "Can't stop aggregation for adu." 4
.IX Item "Can't stop aggregation for adu."
An unrecoverable database error was encountered.  \fBdtpcclock\fR terminates.
.IP "Could not scan outbound Adus" 4
.IX Item "Could not scan outbound Adus"
An unrecoverable database error was encountered.  \fBdtpcclock\fR terminates.
.SH "BUGS"
.IX Header "BUGS"
Report bugs to <ion\-dtn\-support@lists.sourceforge.net>
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fBdtpcadmin\fR\|(1)


release-notes.txt
= Release Notes for ION 4.1.1 =

November 19, 2021

%%%%%%%%%%%
= GENERAL =
%%%%%%%%%%%

The ION (interplanetary overlay network) software is a suite of communication
protocol implementations designed to support mission operation communications
across an end-to-end interplanetary network, which might include on-board
(flight) subnets, in-situ planetary or lunar networks, proximity links, deep
space links, and terrestrial internets.  Included in the ION software
distribution are the following packages:

- ici (interplanetary communication infrastructure), a set of libraries
  that provide flight-software-compatible support for functions on which
  the other packages rely, such as dynamic memory management, non-volatile
  storage management, and inter-task communication via shared memory.
  The ici libraries are designed to make the porting of IPN software to
  multiple operating systems - Linux, VxWorks, Solaris, etc. - as easy as
  possible. Ici now includes zco (zero-copy object), a library that 
  minimizes the copying of application data as it is encapsulated in 
  multiple layers of protocol structure while traversing the protocol
  stack.

- bp (bundle protocol), an implementation of the Delay-Tolerant
  Networking (DTN) architecture's Bundle Protocol.

- ltp (licklider transmission protocol), a DTN convergence layer protocol
  for reliable transmission over links characterized by long or highly
  variable delay.

- dgr (datagram retransmission), an alternative implementation of ltp
  that is designed for use over the Internet protocol stack.  dgr
  implements congestion control and is designed for relatively high
  performance.

- bssp (bundle streaming service protocol), a DTN convergence layer protocol
  for reliable transmission of streaming data.

- ams - an implementation of the CCSDS Asynchronous Message Service.

- cfdp - a class-1 (Unacknowledged) implementation of the CCSDS File
  Delivery Protocol.

- bss - a Bundle Streaming Service (BSS) for disruption-tolerant reliable
  data streaming.  BSS supports real-time streaming applications by
  passing the bundle payloads to the associated application for immediate
  display of the most recent data while storing all bundle payloads
  received into a database for user-directed playback.

- dtpc (delay-tolerant payload conditioning), an application framework
  providing TCP-like support for end-to-end retransmission and data delivery
  in transmission order without duplication or omission.

Features included:

- cgr - Contract graph routing: a method of dynamic routing designed for
  space based applications of ION, but still usable for terrestrial
  applications. It computes routes using scheduled communication and deals
  with time-varying network topology.

- brs - Bundle relay service: provides interconnectivity between networks
  that do not allow servers (those behind NAT for example). For more
  information, check man brsscla and man brsccla.

%%%%%%%%%%%%%%%%%%%%%%%%%%
= NOTES ON RELEASE 4.1.1 =
%%%%%%%%%%%%%%%%%%%%%%%%%%

Beginning with the ION 4.1.1 release, we will no longer be distributing 
two separate .tar.gz files (one that builds with BP version 6 by default and 
one that builds with BP version 7 by default). Instead, we will distribute a
single ION .tar.gz file that builds with BP version 7 by default. If you wish 
to use ION with BP version 6, you can use use the --enable-bpv6 configure 
option. If you are building ION from source on Windows for BP version 6, you
will need to run the command "make bpv6".

After ION version 4, all BP version 6 related changes will be bug fixes
unless explicitly identified as otherwise.

Branches included in this release:

misc-4.1.1-fixes
  * Improve the killm script 
  * Update pilot delay and add rate control for bpdriver
  * Change SourceCustodyRequired to NoCustodyRequested for AMS
  * Add a version of the req-003-multicast test for BP version 7
  * Fix a bug in librams.c that caused AMS petition assert/cancel
    flip-flop
  * Various updates for Makefiles (autoconf and manual)

apl-4.1.1-security-patches
  * Added BPSec Python test suite
  * Added accomodation of NULL terminator in security policy parameters
  * Added checks for NULL pointer references when a BCB targets an 
    extension block
  * Added TEST_POINT macros to BIB and BCB code to log test events

last-pre-git-tweaks
  * Add a warning about potential forwarding loops
  * Fix a typo in biberc documentation

autoconf-updates-MacOS_M1-WSL
  * Fix line ending encoding error in install_macos_sysctl.sh script
  * Add support for building ION on MacOS using the new M1 processor
    * At this time, ION works correctly on the M! processor with MacOS
      version 11.6 (Big Sur) but does not run properly on version
      12.0.1 (Monterey)
  * Add support to configure ION to build on WSL2 but not WSL1
  * ION has also now been tested on RedHat Linux


%%%%%%%%%%%%%%%%%%%%%%%%%%
= NOTES ON RELEASE 4.1.0 =
%%%%%%%%%%%%%%%%%%%%%%%%%%

ION 4.1.0 is the first release in which the BPv7 implementation is
conformant to the BPv7 protocol specification that will shortly be
published by IETF as a standards-track RFC.

The accompanying BPsec implementation is similarly conformant to
the BPSec specification that will also soon be a standard-track RFC.
NOTE, though, that the TCP convergence-layer protocol implementation
in ION 4.1.0 still conforms to the original TCPCL specification (RFC
7242), NOT to the upcoming TCPCL standards-track RFC; that upgrade
will be supported in a future release.

Another key change in this release is that Aggregate Custody Signaling
is no longer enabled by default.  (Due to some new features in ACS
configuration, when ACS is enabled the bpv6 bpclock daemon will no
longer run successfully unless the ACS database is initialized.
Since not all ION bpv6 users currently initialize ACS in node startup,
disabling ACS by default seems to be the best way forward.)  If you

dtpc/daemon/dtpcclock.c
/*
	dtpcclock.c:	Scheduled-event management daemon for DTPC.

	Authors: Giorgos Papastergiou, SPICE
		 Ioannis Alexiadis, SPICE

	Copyright (c) 2011, Space Internetworking Center,
	Democritus University of Thrace. ALL RIGHTS RESERVED.
										*/

#include "dtpcP.h"

static uaddr	_running(uaddr *newValue)
{
	void	*value;
	uaddr	state;
	
	if (newValue)		/*	Changing state.			*/
	{
		value = (void *) (*newValue);
		state = (uaddr) sm_TaskVar(&value);
	}
	else			/*	Just check.			*/
	{
		state = (uaddr) sm_TaskVar(NULL);
	}

	return state;
}

static void	shutDown(int signum)	/*	Stops dtpcclock.	*/
{
	uaddr	stop = 0;

	oK(_running(&stop));	/*	Terminates dtpcclock.		*/
}

static int	updateAdus(Sdr sdr)
{
	DtpcDB		*dtpcConstants = getDtpcConstants();
	Object		aggrElt;
	Object		aduElt;
	Object		profileElt;
	Object		aduObj;
	Object		aggrObj;
	OutAdu		adu;
			OBJ_POINTER(OutAggregator, outAggr);
			OBJ_POINTER(Profile, profile);
	CHKERR(sdr_begin_xn(sdr));
	for (aggrElt = sdr_list_first(sdr, dtpcConstants->outAggregators);
			aggrElt; aggrElt = sdr_list_next(sdr, aggrElt))
	{
		aggrObj = sdr_list_data(sdr, aggrElt);
		GET_OBJ_POINTER(sdr, OutAggregator, outAggr, aggrObj);
		if (outAggr->inProgressAduElt == 0)
		{
			continue;	/*	No ADU in progress.	*/
		}

		aduElt = outAggr->inProgressAduElt;
		aduObj = sdr_list_data(sdr, aduElt);
		sdr_stage(sdr, (char *) &adu, aduObj, sizeof(OutAdu));
		if (adu.ageOfAdu < 0)
		{
			continue;	/*	No items added to ADU.	*/
		}

		adu.ageOfAdu++;
		for (profileElt = sdr_list_first(sdr, dtpcConstants->profiles);
			profileElt; profileElt = sdr_list_next(sdr, profileElt))
		{
			GET_OBJ_POINTER(sdr, Profile, profile,
				sdr_list_data(sdr, profileElt));
			if (outAggr->profileID == profile->profileID)
			{
				break;	/*	Found matching profile.	*/
			}
		}

		sdr_write(sdr, aduObj, (char *) &adu, sizeof(OutAdu));
		if (adu.ageOfAdu >= profile->aggrTimeLimit)
		{
			if (createAdu(profile, aduObj, aduElt) < 0)
			{
				putErrmsg("Can't send finished adu.", NULL);
				sdr_cancel_xn(sdr);
				return -1;
			}

			if (initOutAdu(profile, aggrObj, aggrElt, &aduObj,
					&aduElt) < 0)
			{
				putErrmsg("Can't stop aggregation for adu.",
						NULL);
				sdr_cancel_xn(sdr);
				return -1;
			}
		}
	}

	if (sdr_end_xn(sdr) < 0)
	{
		putErrmsg("Could not scan outbound Adus.", NULL);
		return -1;
	}
	return 0;
}

static int	handleEvents(Sdr sdr, time_t currentTime)
{
	DtpcDB		*dtpcConstants = getDtpcConstants();
	Object		elt;
	Object		eventObj;
	int		result;
			OBJ_POINTER(DtpcEvent, event);

	while (1)
	{
		CHKERR(sdr_begin_xn(sdr));
		elt = sdr_list_first(sdr, dtpcConstants->events);
		if (elt == 0)	/* 	Event list is empty.		*/
		{
			sdr_exit_xn(sdr);
			return 0;
		}	

		eventObj = sdr_list_data(sdr, elt);
		GET_OBJ_POINTER(sdr, DtpcEvent, event, eventObj);
		if (event->scheduledTime > currentTime)
		{
			/*	This is the first future event.		*/

			sdr_exit_xn(sdr);
			return 0;
		}

		switch (event->type)
		{
		case ResendAdu:
			result = resendAdu(sdr, event->aduElt, currentTime);
			break;		/*	Out of switch.		*/

		case DeleteAdu:
			deleteAdu(sdr, event->aduElt);
			if ((getDtpcVdb())->watching & WATCH_m)
			{
				/*
				putchar('m');
				fflush(stdout);
				*/
				iwatch('m');
			}

			result = 0;
			break;		/*	Out of switch.		*/

		case DeleteGap:
			deletePlaceholder(sdr, event->aduElt);
			if ((getDtpcVdb())->watching & WATCH_expire)
			{
				/*
				putchar('*');
				fflush(stdout);
				*/
				iwatch('*');
			}

			result = parseInAdus(sdr);
			break;		/*	Out of switch.		*/

		default:		/*	Spurious event; erase.	*/
			sdr_free(sdr, eventObj);
			sdr_list_delete(sdr, elt, NULL, NULL);
			result = 0;	/*	Event is ignored.	*/
		}

		if (result !=0)
		{
			sdr_cancel_xn(sdr);
			putErrmsg("Failed handling Dtpc event.", NULL);
			return result;
		}

		if (sdr_end_xn(sdr) < 0)
		{
			putErrmsg("Failed handling Dtpc event.", NULL);
			return -1;
		}
	}
}


#if defined (ION_LWT)
int	dtpcclock(saddr a1, saddr a2, saddr a3, saddr a4, saddr a5,
		saddr a6, saddr a7, saddr a8, saddr a9, saddr a10)
{
#else
int	main(int argc, char *argv[])
{
#endif
	Sdr	sdr;
	uaddr	state = 1;
	time_t	currentTime;

	if (dtpcInit() < 0)
	{
		putErrmsg("dtpcclock can't initialize DTPC.", NULL);
		return -1;
	}

	sdr = getIonsdr();
	isignal(SIGTERM, shutDown);

	/* Main loop: wait one second, then parse all events and ADUs.	*/

	oK(_running(&state));
	writeMemo("[i] dtpcclock is running.");

	while (_running(NULL))
	{
		snooze(1);
		currentTime = getCtime();

		/*	Parse list of events				*/

		if (handleEvents(sdr, currentTime) < 0)
		{
			state = 0;	/*	Terminate loop.		*/
			oK(_running(&state));
		}

		/*	Update age for all outbound ADUs		*/

		if (updateAdus(sdr) < 0)
		{
			state = 0;	/*	Terminate loop.		*/
			oK(_running(&state));
		}
	}

	writeErrmsgMemos();
	writeMemo("[i] dtpcclock has ended.");
	ionDetach();
	return 0;
}
